test/buffercomp.c

```diff
@@ -250,6 +250,50 @@ test_multi_keysyms_actions(struct xkb_context *ctx)
 #undef run_test
 }
 
+static void
+test_invalid_symbols_fields(struct xkb_context *ctx)
+{
+    /* Any of the following is invalid syntax, but also use to trigger
+     * a NULL pointer deference, thus this regression test */
+    const char * const keymaps[] = {
+        "xkb_keymap {\n"
+        "    xkb_keycodes { <> = 9; };\n"
+        "    xkb_types { };\n"
+        "    xkb_compat { };\n"
+        "    xkb_symbols { key <> { vmods = [] }; };\n"
```

Comment:
Is `[]` the only way this NULL can happen? The grammar rule is

```y
ArrayInit       :       OBRACKET MultiKeySymList CBRACKET
                        { $$ = $2; }
                |       OBRACKET MultiActionList CBRACKET
                        { $$ = $2; }
                |       OBRACKET CBRACKET
                        { $$ = NULL; }
```

The `OBRACKET CBRACKET` (i.e. `[]`) ends up as NULL, but what if we turn it to an empty `ExprKeysymList` or `ExprActionList` so we don't have to deal with yucky NULL? From my reading of `SetSymbolsField` it will result in a clean "type error" in all cases. And perhaps will allow to remove the `value == NULL` case from `AddSymbolsToKey`.

Reply:
I introduced `EXPR_EMPTY_LIST`, so we have clear semantics: we do not end to check for empty keysym list when processing actions. The first 2 commit requires to be squashed.

I also detected another issue with error handling, see the corresponding commit.

Reply:
> we do not end to check for empty keysym list when processing actions

Right, bummer.

Having a special expression type for empty array is pretty funny but the keysym/action ambiguity forces it.